File: lua/Event.lua
local Event = class("Event")

function Event:Constructor(id)
    self.id = id
    self.handlers = {}
end

---@param handler fun(...): void
---@param context this
function Event:AddListener(handler, context)
    if self.handlers[handler] then
        logwarn("double AddListener ", self.id, handler, context)
    end
    self.handlers[handler] = context or false
end

function Event:RemoveListener(handler)
    if handler then
        if not self.handlers[handler] then
            logwarn("no hander in event ", self.id, handler)
        end
        self.handlers[handler] = nil
    else
        self.handlers = {}
    end
end

function Event:Broadcast(...)
    local count = 0
    for handler, context in pairs(self.handlers) do
        if context then
            handler(context, ...)
        else
            handler(...)
        end
        count = count + 1
    end
    if count == 0 then
        log("event " .. self.id .. " has no listener")
    end
end

return Event

File: src/Event.ts
import { Logger } from "Logger";

export class Event<T> {

    private static teCounter: number = 0;

    private events: Array<{ thisArg: object, handler: (data: T) => void; }>;
    private mId: number;

    constructor() {
        this.mId = Event.teCounter++;
        this.events = [];
    }

    public AddListener(thisArg: object, handler: (data: T | void) => void, once: boolean = false, arg?: T): void {
        for (const elem of this.events) {
            if (elem.thisArg === thisArg && elem.handler === handler) {
                return Logger.Warn(`Event[${this.mId}]: double register`);
            }
        }
        if (once) {
            handler.apply(thisArg, [arg]);
        }
        this.events.push({ thisArg, handler });
    }

    public RemoveListener(thisArg: object, handler: (data: T) => void): void {
        let index: number = -1;
        for (let i = 0; index === -1 && i < this.events.length; i++) {
            const elem = this.events[i];
            if (elem.thisArg === thisArg && elem.handler === handler) {
                index = i;
            }
        }
        if (index === -1) {
            return Logger.Warn(`Event[${this.mId}]: cannot remove`);
        }
        this.events.splice(index, 1);
    }

    public Broadcast(data: T): void {
        if (this.events.length <= 0) {
            return;
        }
        for (const e of this.events) {
            e.handler.apply(e.thisArg, [data]);
        }
    }

}
